web/src/user_status_ui.ts

```diff
@@ -1,19 +1,96 @@
+import type * as tippy from "tippy.js";
 
 import render_set_status_overlay from "../templates/set_status_overlay.hbs";
 import render_status_emoji_selector from "../templates/status_emoji_selector.hbs";
 
 import * as dialog_widget from "./dialog_widget.ts";
+import * as dropdown_widget from "./dropdown_widget.ts";
 import * as emoji from "./emoji.ts";
 import type {EmojiRenderingDetails} from "./emoji.ts";
+import * as flatpickr from "./flatpickr.ts";
 import {$t, $t_html} from "./i18n.ts";
 import * as keydown_util from "./keydown_util.ts";
 import * as people from "./people.ts";
+import * as timerender from "./timerender.ts";
 import * as user_status from "./user_status.ts";
-import type {UserStatusEmojiInfo} from "./user_status.ts";
+import type {TimeKey, UserStatusEmojiInfo} from "./user_status.ts";
 
 let selected_emoji_info: Partial<UserStatusEmojiInfo> = {};
-let default_status_messages_and_emoji_info: {status_text: string; emoji: EmojiRenderingDetails}[];
+let default_status_messages_and_emoji_info: {
+    status_text: string;
+    emoji: EmojiRenderingDetails;
+    expiry: TimeKey;
+}[];
+let scheduled_end_time_widget: dropdown_widget.DropdownWidget;
+let scheduled_end_time: number | null;
+
+type ScheduledEndTimeWidgetOption = {
+    name: string;
+    unique_id: string;
+};
+
+function get_scheduled_end_time_widget_options(): ScheduledEndTimeWidgetOption[] {
+    return [
+        {name: $t({defaultMessage: "Never"}), unique_id: "never"},
+        {name: $t({defaultMessage: "In 30 minutes"}), unique_id: "in_thirty_minutes"},
+        {name: $t({defaultMessage: "In one hour"}), unique_id: "in_one_hour"},
+        {
+            name: $t(
+                {defaultMessage: "Today at {time}"},
+                {
+                    time: timerender.get_localized_date_or_time_for_format(
+                        compute_end_time("today_five_pm")!,
+                        "time",
+                    ),
+                },
+            ),
+            unique_id: "today_five_pm",
+        },
+        {name: $t({defaultMessage: "Tomorrow"}), unique_id: "tomorrow"},
+        {name: $t({defaultMessage: "Custom"}), unique_id: "custom"},
+    ];
+}
+
+const no_scheduled_end_time_set_message = $t({defaultMessage: "Keep status until you remove it."});
+
+function compute_end_time(end_time_string: string, custom_time_selected?: number): number | null {
+    const now = new Date();
+    now.setSeconds(0, 0);
+    const today = new Date(now);
+    const tomorrow = new Date(new Date(now).setDate(now.getDate() + 1));
+
+    switch (end_time_string) {
+        case "never":
+            return null;
+        case "in_thirty_minutes":
+            return new Date(now).setMinutes(now.getMinutes() + 30);
+        case "in_one_hour":
+            return new Date(now).setHours(now.getHours() + 1);
+        case "today_five_pm":
+            return today.setHours(17, 0, 0, 0);
+        case "tomorrow":
+            return tomorrow.setHours(0, 0, 0, 0);
```

Comment:
Have we discussed the decision to have this correspond to midnight in #design?  That seems like a thing that should be considered; a "tomorrow morning" that changes at like 5AM might be more popular.